apps/src/templates/DebugWatch.jsx

```diff
@@ -0,0 +1,425 @@
+    return {
+      text: "",
+      history: [],
+      editing: false,
+      autocompleteOpen: false,
+      autocompleteIndex: 0,
+      autocompleteOptions: DEFAULT_AUTOCOMPLETE_OPTIONS,
+      historyIndex: -1
+    };
+  },
+
+  componentDidMount() {
+    this.wasRunning = null;
+    this.intervalId_ = setInterval(() => {
+      const justStoppedRunning = this.wasRunning && !this.props.isRunning;
+      this.wasRunning = this.props.isRunning;
+
+      if (!this.props.isRunning) {
+        if (justStoppedRunning) {
+          this.props.watchedExpressions.map(we => this.props.update(we.get('expression'), WATCH_VALUE_NOT_RUNNING));
+        }
+        return;
+      }
+
+      this.props.watchedExpressions.map(we => {
+        if (window.tempJSInterpreter) {
+          const currentValue = window.tempJSInterpreter.evaluateWatchExpression(we.get('expression'));
+          this.props.update(we.get('expression'), currentValue);
+        }
+      });
+    }, WATCH_TIMER_PERIOD);
+  },
+
+  componentWillUnmount() {
+    if (this.intervalId_) {
+      clearInterval(this.intervalId_);
+      this.intervalId_ = null;
+    }
+  },
+
+  // http://stackoverflow.com/a/7390612
+  nonValueDescriptor(obj) {
+    return {}.toString.call(obj).split(' ')[1].slice(0, -1).toLowerCase();
+  },
+
+  /**
+   * Gets text to display for given value
+   * @param obj
+   * @returns {*}
+   */
+  renderValue(obj) {
+    const descriptor = this.nonValueDescriptor(obj);
+    const isError = obj instanceof Error;
+
+    if (isError) {
+      return (
+        <span className="watch-value watch-unavailable">
+          {i18n.debugWatchNotAvailable()}
+        </span>
+      );
+    }
+
+    switch (descriptor) {
+      case 'null':
+      case 'undefined':
+        return <span className="watch-value">{descriptor}</span>;
+      case 'regexp':
+        return <span className="watch-value">[regexp]</span>;
+      case 'array':
+        return <span className="watch-value">[array]</span>;
+      case 'function':
+        // [function MyFunctionName]
+        return (
+          <span className="watch-value">
+            {`[${obj.toString().match(/(.*)\(/)[1]}]`}
+          </span>
+        );
+      default:
+        return <span className="watch-value">{obj.toString()}</span>;
+    }
+  },
+
+  scrollToBottom() {
+    this.refs.scrollableContainer.scrollTop = this.refs.scrollableContainer.scrollHeight;
+  },
+
+  addButtonClick() {
+    if (this.state.text === '') {
+      this.setState({
+        autocompleteOpen: true
+      });
+    } else {
+      this.addFromInput();
+    }
+  },
+
+  addFromInput(inputText = this.state.text) {
+    if (inputText === '') {
+      return;
+    }
+    this.props.add(inputText);
+    this.setState({
+      history: this.state.history.concat(inputText),
+      editing: false,
+      text: ''
+    }, () => {
+      this.scrollToBottom();
+      this.filterOptions();
+    });
+  },
+
+  onKeyDown(e) {
+    if (e.key === 'Enter') {
```

Comment:
A bold claim is made: [React normalizes events so that they have consistent properties across different browsers.](https://facebook.github.io/react/docs/events.html#supported-events)